src/dosext/mfs/lfn.c

```diff
@@ -84,28 +84,47 @@ void close_dirhandles(unsigned psp)
 	}
 }
 
-static int vfat_search(char *dest, char *src, char *path, int alias)
+static int vfat_search(char *dest, const char *src, const char *_path, int alias)
 {
-  struct mfs_dir *dir = dos_opendir(path);
+  struct mfs_dir *dir;
   struct mfs_dirent *de;
+  char *path, *slash;
+  int ret = 0;
 
-  if (dir == NULL)
+  d_printf("LFN: vfat_search src=%s path=%s alias=%d\n", src, _path, alias);
+
+  path = strdup(_path);
+  if (!path)
+    return 0;
+
+  slash = strrchr(path, '/');
+  if (!slash) {
+    free(path);
     return 0;
+  }
+  *slash = '\0';
+
+  dir = dos_opendir(path);
+  if (dir == NULL) {
+    free(path);
+    return 0;
+  }
+
   if (dir->dir == NULL)
     while ((de = dos_readdir(dir)) != NULL) {
-      d_printf("LFN: vfat_search %s %s %s %s\n", de->d_name, de->d_long_name, src, path);
+      d_printf("LFN: vfat_search %s %s\n", de->d_name, de->d_long_name);
       if ((strcasecmp(de->d_long_name, src) == 0) || (strcasecmp(de->d_name, src) == 0)) {
-        const char *name = alias ? de->d_name : de->d_long_name;
-        if (!name_ufs_to_dos(dest, name) || alias) {
```

Comment:
Could you please describe in a
log why this is needed? Why that
didn't work and the native one works?

Reply:
Possibly we might want to keep the charset conversion, but converting the filename via mfs/mangle.c is not appropriate for vfat as the mangle scheme is wrong. I don't know if you've used the vfat mounted partition/disk/image under linux (I hadn't), it provides both SFN and LFN for each file, but only LFN is visible from normal `ls`. So you can do:
~~~
total 3
-rwxr-xr-x 1 ajb sphere   9 Oct  6 18:03  MixedCaseFilename.ext
-rwxr-xr-x 1 ajb sphere   9 Oct  6 18:03  shrtname.txt
-rwxr-xr-x 1 ajb sphere   9 Oct  6 18:03 'space embedded filename.txt'
drwxr-xr-x 2 ajb sphere 512 Oct  6 18:03  testname
-rwxr-xr-x 1 ajb sphere   9 Oct  6 18:03  verylongfilename2.txt
-rwxr-xr-x 1 ajb sphere   9 Oct  6 18:03  verylongfilename.txt
~~~

and both names work

~~~
ajb@polly:/clients/common/dosemu2.git$ cat /mnt/dosemu/MixedCaseFilename.ext
Some data
ajb@polly:/clients/common/dosemu2.git$ cat /mnt/dosemu/MIXEDC~1.EXT
Some data
~~~

Which means when looking up a file with SFN it does exist. Then you scan the containing directory looking for a match and you can return the LFN. The opposite LFN -> SFN is done in the same fashion, and there's no need to do any mangling or unmangling.